portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -27,44 +45,135 @@
 public class DataServlet extends HttpServlet {
 
   private final ArrayList<UserComment> userComments = new ArrayList<>();
+  private final String defaultMaxComment = "20";
 
   /*
    * Called when a client submits a GET request to the /data URL
    * Displays all recorded user comments on page
    */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    int maxCommentDisplay = Integer.parseInt(getFieldFromResponse(request, "maxcomments", defaultMaxComment));
+    Query query = new Query("Comment").addSort("time", SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    int displayed = 0;
+    ArrayList<UserComment> comments = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      if(displayed == maxCommentDisplay) {
+          break;
+      }
+      displayed++;
+      long id = entity.getKey().getId();
+      String name = (String) entity.getProperty("name");
+      String email = (String) entity.getProperty("email");
+      String time = (String) entity.getProperty("time");
+      String comment = (String) entity.getProperty("comment");
+      UserComment userComment = UserComment.create(name, email, comment, time);
+      comments.add(userComment);
+    }
+
     Gson gson = new Gson();
-    String json = gson.toJson(userComments);
+
     response.setContentType("application/json;");
-    response.getWriter().println(json);
+    response.getWriter().println(gson.toJson(comments));
   }
 
   /*
    * Called when a client submits a POST request to the /data URL
    * Adds submitted comment to internal record if the comment is
-   * non-empty. Then, the page is reloaded.
+   * non-empty. 
    */
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String userComment = getFieldFromResponse(request, "comment", "");
+    String parsedBody = getBody(request);
+    String trimmedBody = parsedBody.substring(1, parsedBody.length() - 1);
+    Map<String, String> immutableMap = Splitter.on('&').trimResults().withKeyValueSeparator('=').split(trimmedBody);
```

Comment:
Can you use https://www.javadoc.io/doc/com.google.code.gson/gson/latest/com.google.gson/com/google/gson/JsonParser.html to parse it to a Json object?

Reply:
The data is in the form of key1=value1&key2=value2... so it isn't JSON data.

Reply:
Right, but that is because you serialize the form data this way. See my comment below